### PR TITLE
Buffered Transactional Editing

### PR DESCRIPTION
**Pull request by domi4484**
## Buffered Transactional Editing

Implementation of [QGIS-Enhancement-Proposals #203](https://github.com/qgis/QGIS-Enhancement-Proposals/issues/203)
This pr includes the main implementation and follow #46483

Brief summary:
A new "buffered transactions" editing mode for QGIS is added.
With this edit mode, all editable layers are toggled synchronously and all edits are saved in a local edit buffer.
Saving changes is executed within a single transaction on all layers (per provider).

image

In the gif below layers from different providers are simultaneously put into edit mode and persisted.

bufferedTransactions

### CODE REVIEW

File: src/core/vector/qgsvectorlayereditbuffergroup.h [hunk 1]
```diff
@@ -0,0 +1,91 @@
...
+
+class QgsVectorLayer;
+
+/**
+ * \ingroup core
+ * \class QgsVectorLayerEditBufferGroup
+ *
+ * The edit buffer group manages a group of edit buffers. Commands like commit and rollback are
+ * managed by the group invokes individual addFeature(), deleteFeature(), ... in the correct order
+ * across all contained edit buffers.
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayereditbuffergroup.h:32]: since 3.24

File: src/core/vector/qgsvectorlayereditbuffergroup.h [hunk 2]
```diff
@@ -0,0 +1,91 @@
...
+    /**
+     * Gets the list of layers currently managed by this edit buffer group.
+     *
+     * \returns Layer list
+     */
+    QList<QgsVectorLayer *> layers() const;
+
+    QList<QgsVectorLayer *> modifiedLayers() const;
+
+    bool startEditing();
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayereditbuffergroup.h:62]: docstrings

File: src/core/vector/qgsvectorlayereditbuffer.h [hunk 3]
```diff
@@ -198,6 +199,18 @@ class CORE_EXPORT QgsVectorLayerEditBuffer : public QObject
      */
     void updateFields( QgsFields &fields ) SIP_SKIP;
 
+    /**
+     * Returns the parent edit buffer group for this edit buffer, or nullptr if not part of a group.
+     * \since QGIS 3.24
+     */
+    QgsVectorLayerEditBufferGroup *editBufferGroup();
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayereditbuffer.h:206]:
```suggestion
    QgsVectorLayerEditBufferGroup *editBufferGroup() const;
```

File: src/core/vector/qgsvectorlayereditbuffer.cpp [hunk 4]
```diff
@@ -338,53 +348,16 @@ bool QgsVectorLayerEditBuffer::commitChanges( QStringList &commitErrors )
...
-      mChangedGeometries.clear();
-    }
-    else
-    {
-      commitErrors << tr( "ERROR: %n geometries not changed.", "not changed geometries count", mChangedGeometries.size() );
-      success = false;
-    }
-  }
+  if ( success && !mChangedGeometries.isEmpty() )
+    success &= commitChangesUpdateGeometry( commitErrors );
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayereditbuffer.cpp:360]: There is a difference to the proposal here, that geometries are changed as first step, whereas in the proposal they are changed next to attributes. I guess this is the first step for historical reasons and would propose to clean that up and move it to later.

**Review by m-kuhn — COMMENTED** [review]
Thanks, this is looking good.

Questions:
- what happens in case a `startEditing()`, `commitChanges()`, `rollBack()` is called directly on a layer and not through QgsProject?
- can you add some tests to make sure this works as expected when a constraint fails to validate and that always all or no layers are in edit mode after
- (how) is a user notified in case of commit errors?

File: src/core/vector/qgsvectorlayereditbuffer.cpp (hunk 4 above)
Comment (domi4484) [reply src/core/vector/qgsvectorlayereditbuffer.cpp:360]: I moved the update of the geometries after the update of attributes as described in the proposal

File: src/core/vector/qgsvectorlayereditbuffer.h [hunk 5]
```diff
@@ -315,7 +328,23 @@ class CORE_EXPORT QgsVectorLayerEditBuffer : public QObject
...
+
     friend class QgsGrassProvider; //GRASS provider totally abuses the edit buffer
+
+  private:
+
+    friend class QgsVectorLayerEditBufferGroup;
+
+    bool commitChangesCheckGeometryTypeCompatibility( QStringList &commitErrors );
+    bool commitChangesUpdateGeometry( QStringList &commitErrors );
+    bool commitChangesDeleteAttributes( bool &attributesDeleted, QStringList &commitErrors );
```
[the remark this answers is not in the document]
Comment (domi4484) [reply src/core/vector/qgsvectorlayereditbuffer.h:341]: I added some documentation

**Comment by domi4484** [comment]
> * what happens in case a `startEditing()`, `commitChanges()`, `rollBack()` is called directly on a layer and not through QgsProject?
> 

The layer check if the buffered transaction mode is active and calls `startEditing`, `commitChanges` and `rollBack` on `QgsProject`

> * can you add some tests to make sure this works as expected when a constraint fails to validate and that always all or no layers are in edit mode after
> 

Yes I'll add some tests

> * (how) is a user notified in case of commit errors?

As it is now (Message bar + dialog with details)
image

File: src/core/vector/qgsvectorlayer.cpp [hunk 6]
```diff
@@ -1460,6 +1460,9 @@ void QgsVectorLayer::setLabeling( QgsAbstractVectorLayerLabeling *labeling )
 
 bool QgsVectorLayer::startEditing()
 {
+  if ( QgsProject::instance()->transactionMode() == Qgis::TransactionMode::BufferedGroups )
+    return QgsProject::instance()->startEditing( this );
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayer.cpp:1464]: Can we use `project()` here (don't forget a `nullptr` check)?

File: src/core/vector/qgsvectorlayer.cpp [hunk 7]
```diff
@@ -3466,6 +3444,9 @@ QgsFeatureSource::FeatureAvailability QgsVectorLayer::hasFeatures() const
 
 bool QgsVectorLayer::commitChanges( bool stopEditing )
 {
+  if ( QgsProject::instance()->transactionMode() == Qgis::TransactionMode::BufferedGroups )
+    return QgsProject::instance()->commitChanges( mCommitErrors, stopEditing, this );
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayer.cpp:3448]: Best to use it consistently throughout the file

File: src/core/vector/qgsvectorlayer.cpp [hunk 8]
```diff
@@ -3486,7 +3467,13 @@ bool QgsVectorLayer::commitChanges( bool stopEditing )
     return false;
 
   mCommitChangesActive = true;
-  bool success = mEditBuffer->commitChanges( mCommitErrors );
+
+  bool success = false;
+  if ( mEditBuffer->editBufferGroup() )
```
Comment (m-kuhn) [inline src/core/vector/qgsvectorlayer.cpp:3472]: Would it make sense to move this block into `QgsEditBuffer::commitChanges()`? That would reduce the requirement to write it whenever calling `editBuffer.commitChanges()` (and potentially forgetting it).

**Review by m-kuhn — APPROVED** [review]
Looking very good, thanks.

File: src/core/vector/qgsvectorlayereditbuffergroup.h [hunk 9]
```diff
@@ -0,0 +1,116 @@
...
+    /**
+     * Returns TRUE if the layers are in editing mode
+     */
+    bool isEditing() const;
+
+  private:
+
+    QList<QgsVectorLayer *> orderLayersParentsToChildren( QList<QgsVectorLayer *> layers );
+
+    QList<QgsVectorLayer *> mLayers;
```
[the remark this answers is not in the document]
Comment (domi4484) [reply src/core/vector/qgsvectorlayereditbuffergroup.h]: Yes I was already thinking of changing it to QSet... know it's done

File: src/core/project/qgsproject.cpp [hunk 10]
```diff
@@ -627,6 +627,68 @@ QgsPropertyCollection QgsProject::dataDefinedServerProperties() const
...
+}
+
+bool QgsProject::rollBack( bool stopEditing, QgsVectorLayer *vectorLayer )
+{
+  switch ( mTransactionMode )
+  {
+    case Qgis::TransactionMode::Disabled:
+    case Qgis::TransactionMode::AutomaticGroups:
+    {
+      if ( ! vectorLayer )
```
[the remark this answers is not in the document]
Comment (domi4484) [reply src/core/project/qgsproject.cpp:682]: Yes, I also added in general more detailed information about rollback errors

File: src/core/project/qgsproject.h [hunk 11]
```diff
@@ -2047,7 +2047,7 @@ class CORE_EXPORT QgsProject : public QObject, public QgsExpressionContextGenera
      *
      * \since QGIS 3.26
      */
-    bool rollBack( bool stopEditing = true, QgsVectorLayer *vectorLayer = nullptr );
+    bool rollBack( QStringList &rollbackErrors, bool stopEditing = true, QgsVectorLayer *vectorLayer = nullptr );
```
Comment (m-kuhn) [inline src/core/project/qgsproject.h:2050]: Should this be `SIP_OUT` (that will add it to the return value in python)?

File: src/core/vector/qgsvectorlayer.cpp (hunk 7 above)
Comment (domi4484) [reply src/core/vector/qgsvectorlayer.cpp]: I have changed it everywhere i used it for this PR. In numerous other places `QgsVectorLayer` uses `QgsProject::instance()`. I think all this occurrences should also be updated... I could do it in a separate pr

File: src/core/project/qgsproject.h (hunk 11 above)
Comment (domi4484) [reply src/core/project/qgsproject.h:2050]: Yes it would make sense, i was not aware of this

File: src/core/vector/qgsvectorlayer.cpp (hunk 7 above)
Comment (m-kuhn) [reply src/core/vector/qgsvectorlayer.cpp]: Yes, please do, IIRC I cleaned that up a year ago but there is no way to check that in a future proof way.

File: src/core/vector/qgsvectorlayer.cpp (hunk 8 above)
Comment (domi4484) [reply src/core/vector/qgsvectorlayer.cpp:3472]: Mmm then i'd had to pass also `stopEditing` to `QgsEditBuffer::commitChanges()` which doesn't have any other usage for it than pass it to `QgsEditBufferGroup`. The logic of stop editing is matter of the layer.

Comment (m-kuhn) [reply src/core/vector/qgsvectorlayer.cpp:3472]: What happens currently if `editBuffer.commitChanges()` is called when the mode is buffered?

Comment (domi4484) [reply src/core/vector/qgsvectorlayer.cpp:3472]: The changes of that buffer will be committed. But does it make sense to call that function from other places than `QgsVectorLayer::commitChanges`?
I feel the decision (commit over group/single buffer) should stay in `QgsVectorLayer`. Then `QgsEditBuffer` does not even need to know that he his part of a group (`mEditBufferGroup` could be a direct member of `QgsVectorLayer`)

**Comment by domi4484** [comment]
Could this be merged to get some testing before next release? MingW build seems to be generally broken and not related to this pr

**Comment by nyalldawson** [comment]
@domi4484 
there's an annoying regression since this was merged -- now saving unchanged edits on a layer immediately makes that layer non-editable. The layer should remain editable after saving.

**Comment by domi4484** [comment]
@nyalldawson I'll take a look today

**Comment by agiudiceandrea** [comment]
Please have a look at the related issue report: https://github.com/qgis/QGIS/issues/59828.

**Pull request merged**